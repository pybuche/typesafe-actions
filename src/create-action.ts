import {
  TypeConstant,
  ActionCreatorBuilder,
  ActionCreatorTypeMetadata,
  ActionBuilder,
} from './type-helpers';
import { createCustomAction } from './create-custom-action';
import {
  checkIsEmpty,
  throwIsEmpty,
  checkInvalidActionType,
  throwInvalidActionType,
} from './utils/validation';

/**
 * @description create an action-creator of a given function that contains hidden "type" metadata
 */
export function createAction<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined,
  TArgs extends any[] = any[]
>(
  type: TType,
  payloadCreator: undefined | ((...args: TArgs) => TPayload),
  metaCreator?: (...args: TArgs) => TMeta
): () => (...args: TArgs) => ActionBuilder<TType, TPayload, TMeta>;

export function createAction<TType extends TypeConstant>(
  type: TType
): <TPayload = undefined, TMeta = undefined>() => ActionCreatorBuilder<
  TType,
  TPayload,
  TMeta
>;

export function createAction<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined,
  TArgs extends any[] = any[]
>(
  type: TType,
  payloadCreator?: undefined | ((...args: TArgs) => TPayload),
  metaCreator?: (...args: TArgs) => TMeta
):
  | (() => (...args: TArgs) => ActionBuilder<TType, TPayload, TMeta>)
  | (<TPayload = undefined, TMeta = undefined>() => ActionCreatorBuilder<
      TType,
      TPayload,
      TMeta
    >) {
  if (checkIsEmpty(type)) {
    throwIsEmpty(1);
  }

  if (checkInvalidActionType(type)) {
    throwInvalidActionType(1);
  }

  if (payloadCreator != null || metaCreator != null) {
    return <TPayload, TMeta = undefined>() => {
      return createCustomAction(type, (...args: TArgs) => {
        const payload =
          payloadCreator != null ? payloadCreator(...args) : undefined;
        const meta = metaCreator != null ? metaCreator(...args) : undefined;

        return {
          ...(payload !== undefined && { payload }),
          ...(meta !== undefined && { meta }),
        };
      }) as ActionCreatorBuilder<TType, TPayload, TMeta>;
    };
  }

  return <TPayload, TMeta = undefined>() => {
    return createCustomAction(type, (...args: TArgs) => {
      const payload = args[0];
      const meta = args[1];

      return {
        ...(payload !== undefined && { payload }),
        ...(meta !== undefined && { meta }),
      };
    }) as ActionCreatorBuilder<TType, TPayload, TMeta>;
  };
}
