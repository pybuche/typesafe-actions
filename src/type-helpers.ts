/**
 * PUBLIC API
 */

/**
 * @desc Interface for internal types augmentation
 * @example ```
 * declare module 'typesafe-actions' {
 *   export type RootAction = ActionType<typeof import('./root-action').default>;
 *   export interface Types {
 *     RootAction: RootAction;
 *   }
 * } ```
 */
export interface Types {}

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant = TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

// /**
//  * @desc Action with only Meta
//  */
// export type MetaAction<TType extends TypeConstant, TMeta> = {
//   type: TType;
//   meta: TMeta;
// };

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyActionCreator<TType extends TypeConstant> = () => EmptyAction<
  TType
>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadActionCreator<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

// /**
//  * @desc Action Creator producing MetaAction
//  */
// export type MetaActionCreator<TType extends TypeConstant, TMeta> = (
//   payload: undefined,
//   meta: TMeta
// ) => MetaAction<TType, TMeta>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaActionCreator<
  TType extends TypeConstant,
  TPayload,
  TMeta
> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface ActionCreatorTypeMetadata<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>;
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionCreatorBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadActionCreator<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaActionCreator<TType, TPayload, TMeta>
      : EmptyActionCreator<TType>
    : PayloadActionCreator<TType, TPayload>
  : PayloadMetaActionCreator<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilder<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAction<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAction<TType, TPayload, TMeta>
      : EmptyAction<TType>
    : PayloadAction<TType, TPayload>
  : PayloadMetaAction<TType, TPayload, TMeta>;

/** @private */
export type ResolveType<T extends any> = T extends Function
  ? T
  : { [K in keyof T]: T[K] };
