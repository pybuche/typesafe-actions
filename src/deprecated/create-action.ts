import { TypeConstant, ActionCreator } from '../type-helpers';
import { action } from '../action';

export type PayloadMetaAction<
  T extends TypeConstant,
  P,
  M
> = P extends undefined
  ? M extends undefined
    ? { type: T }
    : { type: T; meta: M }
  : M extends undefined
  ? { type: T; payload: P }
  : { type: T; payload: P; meta: M };

/**
 * @description typesafe action-creator factory
 */
export function createAction<
  T extends TypeConstant,
  AC extends ActionCreator<T> = () => { type: T }
>(
  type: T,
  createHandler?: (
    actionCallback: <P = undefined, M = undefined>(
      payload?: P,
      meta?: M
    ) => PayloadMetaAction<T, P, M>
  ) => AC
): AC {
  // validation is already done in action function

  const actionCreator: AC =
    createHandler == null
      ? ((() => action(type)) as AC)
      : createHandler(action.bind(null, type) as Parameters<
          typeof createHandler
        >[0]);

  return Object.assign(actionCreator, {
    getType: () => type,
    // redux-actions compatibility
    toString: () => type,
  });
}
