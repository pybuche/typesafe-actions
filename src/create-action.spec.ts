import * as TH from './type-helpers';
import { createAction } from './create-action';

describe('Typesafe Actions', () => {
  describe('toString() method return a type', () => {
    const actionCreator = createAction('CREATE_ACTION')();
    // @dts-jest:pass:snap
    actionCreator.toString(); // => 'CREATE_ACTION'
  });

  describe('with symbol', () => {
    const CREATE_ACTION = Symbol(1);
    const withSymbol = createAction(CREATE_ACTION as any)();
    // @dts-jest:pass:snap
    withSymbol(); // => { type: CREATE_ACTION }
  });

  describe('with type only - undefined', () => {
    const withTypeOnly = createAction('CREATE_ACTION')<undefined>();
    // @dts-jest:pass:snap
    withTypeOnly(); // => { type: 'CREATE_ACTION' }
  });

  describe('with payload - number', () => {
    const withPayload = createAction('CREATE_ACTION')<number>();
    // @dts-jest:pass:snap
    withPayload(10); // => { type: 'CREATE_ACTION', payload: 10 }
  });

  describe('with payload - boolean', () => {
    const withPayload = createAction('CREATE_ACTION')<boolean>();
    // @dts-jest:pass:snap
    withPayload(true); // => { type: 'CREATE_ACTION', payload: true }
  });

  describe('with payload - literal string union', () => {
    type NetStatus = 'up' | 'down' | 'unknown';
    const withPayload = createAction('CREATE_ACTION')<NetStatus>();
    // @dts-jest:pass:snap
    withPayload('up'); // => { type: 'CREATE_ACTION', payload: 'up' }
  });

  describe('with payload - primitives union', () => {
    const withPayload = createAction('CREATE_ACTION')<string | null | number>();
    // @dts-jest:pass:snap
    withPayload('foo'); // => { type: 'CREATE_ACTION', payload: 'foo' }
    // @dts-jest:pass:snap
    withPayload(null); // => { type: 'CREATE_ACTION', payload: null }
    // @dts-jest:pass:snap
    withPayload(3); // => { type: 'CREATE_ACTION', payload: 3 }
  });

  describe('with payload - any', () => {
    const withPayload = createAction('CREATE_ACTION')<any>();
    // @dts-jest:pass:snap
    withPayload(10); // => { type: 'CREATE_ACTION', payload: 10 }
  });

  describe('with meta', () => {
    const withMeta = createAction('CREATE_ACTION')<undefined, string>();
    // @dts-jest:pass:snap
    withMeta(undefined, 'token'); // => { type: 'CREATE_ACTION', meta: 'token' }
  });

  describe('with meta - any', () => {
    const withMeta = createAction('CREATE_ACTION')<undefined, any>();
    // @dts-jest:pass:snap
    withMeta(undefined, 'token'); // => { type: 'CREATE_ACTION', meta: 'token' }
  });

  describe('with payload and meta', () => {
    const withPayloadAndMeta = createAction('CREATE_ACTION')<number, string>();
    // @dts-jest:pass:snap
    withPayloadAndMeta(1, 'token'); // => { type: 'CREATE_ACTION', payload: 1, meta: 'token' }
  });

  describe('type-safe usage with higher-order function', () => {
    interface UserSettingsState {
      settingA: string;
      settingB: number;
    }

    const setUserSetting = <K extends keyof UserSettingsState>(
      setting: K,
      newValue: UserSettingsState[K]
    ) =>
      createAction('SET_USER_SETTING')<{
        setting: typeof setting;
        newValue: typeof newValue;
      }>()({ setting, newValue });

    // @dts-jest:pass:snap
    setUserSetting('settingA', 'foo');
    // @dts-jest:fail:snap
    setUserSetting('settingA', 0); // Error as expected

    // @dts-jest:pass:snap
    setUserSetting('settingB', 0);
    // @dts-jest:fail:snap
    setUserSetting('settingB', 'foo'); // Error as expected
  });
});

describe('Redux Actions', () => {
  describe('with type only', () => {
    const withTypeOnly = createAction('CREATE_ACTION')();
    // @dts-jest:pass:snap
    withTypeOnly(); // => { type: 'CREATE_ACTION' }
  });

  describe('with payload - primitive param', () => {
    const withPayload = createAction(
      'CREATE_ACTION',
      (payload: string) => payload
    )();
    // @dts-jest:pass:snap
    withPayload('info message'); // => { type: 'CREATE_ACTION', payload: 'info message' }
  });

  describe('with payload - union param', () => {
    const withPayload = createAction(
      'CREATE_ACTION',
      (payload: string | null | number) => payload
    )();
    // @dts-jest:pass:snap
    withPayload('info message'); // => { type: 'CREATE_ACTION', payload: 'info message' }
    // @dts-jest:pass:snap
    withPayload(null); // => { type: 'CREATE_ACTION', payload: null }
    // @dts-jest:pass:snap
    withPayload(3); // => { type: 'CREATE_ACTION', payload: 3 }
  });

  describe('with meta - primitive param', () => {
    const withMeta = createAction(
      'CREATE_ACTION',
      undefined,
      (meta: string) => meta
    )();
    // @dts-jest:pass:snap
    withMeta('info message'); // => { type: 'CREATE_ACTION', meta: 'info message' }
  });

  describe('with meta - union param', () => {
    const withMeta = createAction(
      'CREATE_ACTION',
      undefined,
      (meta: string | null | number) => meta
    )();
    // @dts-jest:pass:snap
    withMeta('info message'); // => { type: 'CREATE_ACTION', meta: 'info message' }
    // @dts-jest:pass:snap
    withMeta(null); // => { type: 'CREATE_ACTION', meta: null }
    // @dts-jest:pass:snap
    withMeta(3); // => { type: 'CREATE_ACTION', meta: 3 }
  });

  describe('with payload and meta - primitive params', () => {
    const withPayloadAndMeta = createAction(
      'CREATE_ACTION',
      (username: string, message: string) => `${username}: ${message}`,

      (username: string, message: string) => ({
        username,
        message,
      })
    )();
    // tslint:disable:max-line-length
    // @dts-jest:pass:snap
    withPayloadAndMeta('Piotr', 'Hello!'); // => { type: 'CREATE_ACTION', payload: 'Piotr: Hello!', meta: { username: 'Piotr', message: 'Hello!' } }
  });
});
