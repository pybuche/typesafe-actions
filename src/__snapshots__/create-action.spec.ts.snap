// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionCreator.toString() 1`] = `"string"`;

exports[`setUserSetting('settingA', 'foo') 1`] = `"TH.PayloadAction<\\"SET_USER_SETTING\\", { setting: \\"settingA\\"; newValue: string; }>"`;

exports[`setUserSetting('settingA', 0) 1`] = `"Argument of type '0' is not assignable to parameter of type 'string'."`;

exports[`setUserSetting('settingB', 'foo') 1`] = `"Argument of type '\\"foo\\"' is not assignable to parameter of type 'number'."`;

exports[`setUserSetting('settingB', 0) 1`] = `"TH.PayloadAction<\\"SET_USER_SETTING\\", { setting: \\"settingB\\"; newValue: number; }>"`;

exports[`withMeta('info message') 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string>"`;

exports[`withMeta('info message') 2`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`withMeta(3) 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`withMeta(null) 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string | number | null>"`;

exports[`withMeta(undefined, 'token') 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, string>"`;

exports[`withMeta(undefined, 'token') 2`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", undefined, any>"`;

exports[`withPayload('foo') 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload('info message') 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string>"`;

exports[`withPayload('info message') 2`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload('up') 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", \\"up\\" | \\"down\\" | \\"unknown\\">"`;

exports[`withPayload(3) 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload(3) 2`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload(10) 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", number>"`;

exports[`withPayload(10) 2`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", any>"`;

exports[`withPayload(null) 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload(null) 2`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", string | number | null>"`;

exports[`withPayload(true) 1`] = `"TH.PayloadAction<\\"CREATE_ACTION\\", boolean>"`;

exports[`withPayloadAndMeta('Piotr', 'Hello!') 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", string, { username: string; message: string; }>"`;

exports[`withPayloadAndMeta(1, 'token') 1`] = `"TH.PayloadMetaAction<\\"CREATE_ACTION\\", number, string>"`;

exports[`withSymbol() 1`] = `"TH.EmptyAction<any>"`;

exports[`withTypeOnly() 1`] = `"TH.EmptyAction<\\"CREATE_ACTION\\">"`;

exports[`withTypeOnly() 2`] = `"TH.EmptyAction<\\"CREATE_ACTION\\">"`;
